client/concordclient/include/client/concordclient/trc_queue.hpp

```diff
@@ -0,0 +1,112 @@
+// Concord
+//
+// Copyright (c) 2020-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <condition_variable>
+#include <boost/lockfree/spsc_queue.hpp>
+#include <list>
+#include <memory>
+#include <thread>
+#include <mutex>
+#include <exception>
+#include <chrono>
+#include <variant>
+
+#include "assertUtils.hpp"
+#include "event_update.hpp"
+
+namespace concord::client::concordclient {
+
+typedef std::variant<Update, EventGroup> EventVariant;
+
+class TrcQueue {
+ private:
+  boost::lockfree::spsc_queue<EventVariant*> queue_data_;
+  std::exception_ptr exception_;
+  std::mutex mutex_;
```

Comment:
if we are using the lockfree queue, then we should remove these mutex and condition variable.

Reply:
We use mutex and condition variable when popping from the queue if the queue is empty, and in `setException`.

Reply:
ok, so the mutex is for exception_ variable. thanks.